### PR TITLE
Isotropic_ remeshing_plugin use SM

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes

The isotropic_remeshing_plugin can work with Scene_surface_mesh_items
 
This PR is based on #1983
## Release Management

* Affected package(s):Polyhedron_demo
* Issue(s) solved (if any): fix #1931

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+// -----------------------------------
+// Private methods
+// -----------------------------------
+template <typename P_, typename I_>
+typename Detect_features_in_polyhedra<P_, I_>::Vector_3
+Detect_features_in_polyhedra<P_, I_>::
+facet_normal(const Polyhedron& polyhedron, const face_descriptor& f) const
+{
+  return Polygon_mesh_processing::compute_face_normal(f,polyhedron);
+  /*
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h]: why is this code commented? should it be removed?

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    return true;
+  else
+    return false;
+}
+  
+ 
+template <typename P_, typename I_>
+void
+Detect_features_in_polyhedra<P_, I_>::
+flood(Polyhedron& polyhedron,face_descriptor f, const Patch_id patch_id, face_descriptor_set& unsorted_faces) const
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h]: please replace `const Patch_id patch_id` with `const Patch_id& patch_id`

Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h]: and also split this line into 3. It's too long

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    if ( ! he->is_feature_edge() && ! is_border(he,polyhedron) )
+    {
+      face_descriptor explored_facet = face(he,polyhedron);
+      
+      // Mark facet and delete it from unsorted
+      put(pid_map, explored_facet, patch_id);
+      unsorted_faces.erase(explored_facet);
+      
+      // Add/Remove facet's halfedge to/from explore list
+      BOOST_FOREACH(halfedge_descriptor hd, halfedges_around_face(halfedge(explored_facet,polyhedron), polyhedron))
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h]: split this line into 2 please

File: Polyhedron/demo/Polyhedron/Plugins/IO/Selection_io_plugin.cpp [hunk 4]
```diff
@@ -10,7 +10,12 @@ class Polyhedron_demo_selection_io_plugin :
     Q_INTERFACES(CGAL::Three::Polyhedron_demo_io_plugin_interface)
     Q_PLUGIN_METADATA(IID "com.geometryfactory.PolyhedronDemo.PluginInterface/1.0")
 public:
+#ifdef USE_SURFACE_MESH
+    QString name() const { return "selection_io_sm_plugin"; }
+#else
```
Comment (janetournois) [inline Polyhedron/demo/Polyhedron/Plugins/IO/Selection_io_plugin.cpp]: Is this suited? (it looks that the ifdef was belonging to a preliminary version)

File: Polyhedron/demo/Polyhedron/Plugins/Mesh_3/Detect_sharp_edges_plugin.cpp [hunk 5]
```diff
@@ -105,15 +105,15 @@ void Polyhedron_demo_detect_sharp_edges_plugin::detectSharpEdges(bool input_dial
...
-  CGAL::Mesh_3::Detect_features_in_polyhedra<Polyhedron> detect_features;
+  typedef boost::property_map<Polyhedron,CGAL::face_patch_id_t>::type PatchID;
   Q_FOREACH(Poly_tuple tuple, polyhedrons)
   {
     Scene_polyhedron_item* item =
       qobject_cast<Scene_polyhedron_item*>(scene->item(tuple.first));
     Polyhedron* pMesh = tuple.second;
     if (!pMesh) continue;
 
+    CGAL::Polygon_mesh_processing::Detect_features_in_polyhedra<Polyhedron, PatchID> detect_features(get(CGAL::face_patch_id_t(),*pMesh));
```
Comment (janetournois) [inline Polyhedron/demo/Polyhedron/Plugins/Mesh_3/Detect_sharp_edges_plugin.cpp]: this line is too long
please split it

**Comment by janetournois** [comment]
I understand, from this PR, that the strategy is to discard `Polyhedron_3` from the `isotropic_remeshing_plugin`, as soon as `Surface_mesh` is available. Am I right? 
If so, is it suited?

**Comment by maxGimeno** [comment]
I think that your whole review concerns other branches, on which this one is based. See #1954 and #1983 .

The Strategy is to make all the plugins work for both Polyhderon and Surface_mesh for an undetermined time, at least until everything works with Surface_mesh, which may take a little while.

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item_k_ring_selection.h [hunk 6]
```diff
@@ -146,6 +189,33 @@ public Q_SLOTS:
     updateIsTreated();
   }
 
+  // slots are called by signals of surface_mesh_item
+  void sm_vertex_has_been_selected(void* v)
+  {
+    std::size_t h = reinterpret_cast<std::size_t>(v);
+    is_active=true;
+    if(active_handle_type == Active_handle::VERTEX || active_handle_type == Active_handle::PATH)
+      process_selection( static_cast<sm_vertex_descriptor>(h) );
```
Comment (janetournois) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item_k_ring_selection.h]: Sorry if this is part of PR #1983 but I compiled this branch. This line causes a conversion warning from `std::size_t` to `CGAL::SM_Index<CGAL::SM_Vertex_index>::size_type` here

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item_k_ring_selection.h [hunk 7]
```diff
@@ -146,6 +189,33 @@ public Q_SLOTS:
...
+    updateIsTreated();
+  }
+  void sm_facet_has_been_selected(void* v)
+  {
+    std::size_t h = reinterpret_cast<std::size_t>(v);
+    is_active=true;
+    if (active_handle_type == Active_handle::FACET
+      || active_handle_type == Active_handle::CONNECTED_COMPONENT)
+      process_selection(static_cast<sm_face_descriptor>(h) );
```
Comment (janetournois) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item_k_ring_selection.h]: conversion warning here too : from `size_t` to  CGAL::SM_Index<CGAL::SM_Face_index>::size_type`

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item_k_ring_selection.h [hunk 8]
```diff
@@ -146,6 +189,33 @@ public Q_SLOTS:
...
+    updateIsTreated();
+  }
+  void sm_edge_has_been_selected(void* v)
+  {
+    std::size_t h = reinterpret_cast<std::size_t>(v);
+    is_active=true;
+    if(active_handle_type == Active_handle::EDGE)
+      process_selection(static_cast<sm_edge_descriptor>(h) );
```
Comment (janetournois) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item_k_ring_selection.h]: conversion warning from `size_t` to `CGAL::SM_Edge_index::size_type`

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_selection_item.h [hunk 9]
```diff
@@ -661,73 +701,73 @@ friend class Polyhedron_demo_selection_plugin;
...
-    return out->size_of_vertices() > 0;
+    return num_vertices(*out) > 0;
   }
 
   void select_sharp_edges(const double angle)
   {
     CGAL::detect_sharp_edges(polyhedron(), angle);
 
+    CGAL::halfedge_is_feature_t hif;
+    boost::property_map<Face_graph,CGAL::halfedge_is_feature_t>::type is_feature = get(hif,*polyhedron());
```
Comment (janetournois) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_selection_item.h]: warning : `hif` is used but not initialized

File: Polyhedron/demo/Polyhedron/Plugins/PMP/Isotropic_remeshing_plugin.cpp [hunk 10]
```diff
@@ -192,52 +218,39 @@ class Polyhedron_demo_isotropic_remeshing_plugin :
...
+                                   std::map<FaceGraph*,Edge_set>& edges_to_protect,
                                    double target_length)
   {
     typedef Polyhedron::Point_3 Point_3;
     typedef std::pair<Point_3,Point_3> Segment_3;
 
     typedef std::map< Segment_3,
-                      std::vector< std::pair<halfedge_descriptor, Polyhedron*> > > MapType;
+                      std::vector< std::pair<halfedge_descriptor, FaceGraph*> > > MapType;
+    typedef typename boost::property_map<FaceGraph,
```
Comment (janetournois) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/Isotropic_remeshing_plugin.cpp]: compilation error : `typename` cannot be used outside of a template class

File: Polyhedron/demo/Polyhedron/Plugins/IO/Selection_io_plugin.cpp (hunk 4 above)
Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/Plugins/IO/Selection_io_plugin.cpp]: No, it is intended. In the CMakeLists.txt we define two different targets linked with different libs, it seems legit to not name them the same, IMHO.

**Comment by maxGimeno** [comment]
Ready for a new review.

**Pull request merged**